### PR TITLE
Implement subset minimal unsat core

### PR DESCRIPTION
**Pull request by Tomaqa**

### CODE REVIEW

File: src/api/Interpret.cc [hunk 1]
```diff
@@ -1285,14 +1285,24 @@ void Interpret::getProof()
...
+    } else {
+        for (PTRef fla : unsatCore->getTerms()) {
+            if (termNames.contains(fla)) {
+                std::cout << termNames.nameForTerm(fla);
+            }
+            else {
+                std::cout << logic->printTerm(fla);
+            }
+            std::cout << ' ';
+        }
```
Comment (blishko) [inline src/api/Interpret.cc:1305]: Why is this needed?

Comment (Tomaqa) [reply src/api/Interpret.cc:1305]: Just (slight) performance reasons. In the first branch, we use named terms so all of them must be in `termNames`. In second branch, we iterate over all terms so not all of them are necessarily in `termNames`. If it is not, we just print out string representation of the term, because we do not have any name for it.

Comment (Tomaqa) [reply src/api/Interpret.cc:1305]: This commit implements option `:print-cores-full` from https://cvc5.github.io/blog/2024/04/15/interfaces-for-understanding-cvc5.html

Comment (blishko) [reply src/api/Interpret.cc:1305]: I see, can you please document the meaning of this option? And maybe also the `minimal-unsat-cores`, to make it clear that these are subset minimal, and not global minimum.

File: test/regression/base/generic/trivial_unsat_core2_full.smt2.expected.out [hunk 2]
```diff
@@ -0,0 +1,2 @@
+unsat
+( b1 a3 )
```
Comment (blishko) [inline test/regression/base/generic/trivial_unsat_core2_full.smt2.expected.out:2]: This is not correct. We should not mix terms and name together.
`cvc5` returns 
```
unsat
(
b1
(not b1)
)
```

**Review by blishko — CHANGES_REQUESTED** [review]

File: test/regression/base/QF_UF/php_3_3_40_min_unsat_core.smt2
Comment (blishko) [inline test/regression/base/QF_UF/php_3_3_40_min_unsat_core.smt2:1]: Do we need these large files? I would prefer to minimize as much as possible the regression tests that we add.

File: test/regression/base/generic/trivial_unsat_core2_full.smt2.expected.out (hunk 2 above)
Comment (Tomaqa) [reply test/regression/base/generic/trivial_unsat_core2_full.smt2.expected.out:2]: Sounds reasonable.

File: test/regression/base/QF_UF/php_3_3_40_min_unsat_core.smt2
Comment (Tomaqa) [reply test/regression/base/QF_UF/php_3_3_40_min_unsat_core.smt2:1]: I removed them from this PR. However, I think some of them will be needed to track the minimization process more thoroughly. They could be simplified, though.

Comment (Tomaqa) [reply test/regression/base/QF_UF/php_3_3_40_min_unsat_core.smt2:1]: -> #770

File: src/api/MainSolver.cc [hunk 3]
```diff
@@ -231,9 +234,15 @@ std::unique_ptr<UnsatCore> MainSolver::getUnsatCore() const {
     if (not config.produce_unsat_cores()) { throw ApiException("Producing unsat cores is not enabled"); }
     if (status != s_False) { throw ApiException("Unsat core cannot be extracted if solver is not in UNSAT state"); }
 
-    UnsatCoreBuilder unsatCoreBuilder{&config, &smt_solver->getResolutionProof(), &pmanager};
+    auto & proof = smt_solver->getResolutionProof();
+    auto unsatCoreBuilder = [&]() -> std::variant<UnsatCoreBuilder, MinUnsatCoreBuilder> {
+        if (not config.minimal_unsat_cores()) { return UnsatCoreBuilder{config, proof, pmanager, termNames}; }
+
+        auto auxSMTSolver = std::make_unique<MainSolver>(logic, config, "unsat core solver");
```
Comment (blishko) [inline src/api/MainSolver.cc:241]: It might be more efficient to give this new solver a fresh config. Specifically, the new solver does not need to compute unsat cores anymore.

File: src/unsatcores/MinUnsatCoreBuilder.cc [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    minimizeAlg();
+    minimizeFinish();
+}
+
+void MinUnsatCoreBuilder::minimizeInit() {
+    assert(terms.size() >= namedTerms.size());
+    assert(size_t(namedTerms.size()) == namedTermsIdxs.size());
+}
+
+void MinUnsatCoreBuilder::minimizeAlgNaive() {
```
Comment (blishko) [inline src/unsatcores/MinUnsatCoreBuilder.cc:29]: This algorithm only minimizes the named terms, no?
Thus this does not compute minimal explanation if used together with the option `print-cores-full`, no?

**Review by blishko — COMMENTED** [review]
Looks good!
But it seems the two options that you are introducing does not work together correctly.
Maybe the full cores should be introduced in a separate PR? Since it is orthogonal functionality.

Another note is that we should investigate using the [Quickxplain algorithm](https://cdn.aaai.org/AAAI/2004/AAAI04-027.pdf) to compute minimal unsat cores.

File: src/api/MainSolver.cc (hunk 3 above)
Comment (Tomaqa) [reply src/api/MainSolver.cc:241]: That is a good point.

File: src/unsatcores/MinUnsatCoreBuilder.cc (hunk 4 above)
Comment (blishko) [reply src/unsatcores/MinUnsatCoreBuilder.cc:29]: My intuition is that if `print-cores-full` is set then the names should be ignored and, if minimal cores are required then the minimization algorithm should operate on all terms, not only the named ones.

File: src/api/Interpret.cc (hunk 1 above)
Comment (Tomaqa) [reply src/api/Interpret.cc]: Moved to #771

File: src/api/MainSolver.cc (hunk 3 above)
Comment (Tomaqa) [reply src/api/MainSolver.cc:241]: Based on this, I actually reconsidered the design and I think both builder classes can be merged. Since we use `MainSolver` and not `SimpSMTSolver`, it is sufficient to extend `UnsatCoreBuilder` constructor of just `Logic &` parameter, which is fine. The internal SMT solver can be constructed inside only if minimization is desired. Do you agree?

Comment (blishko) [reply src/api/MainSolver.cc:241]: That would be very good, I think!

Comment (Tomaqa) [reply src/api/MainSolver.cc]: Done.

File: src/api/MainSolver.cc [hunk 5]
```diff
@@ -231,7 +231,8 @@ std::unique_ptr<UnsatCore> MainSolver::getUnsatCore() const {
     if (not config.produce_unsat_cores()) { throw ApiException("Producing unsat cores is not enabled"); }
     if (status != s_False) { throw ApiException("Unsat core cannot be extracted if solver is not in UNSAT state"); }
 
-    UnsatCoreBuilder unsatCoreBuilder{&config, &smt_solver->getResolutionProof(), &pmanager};
+    auto & proof = smt_solver->getResolutionProof();
+    UnsatCoreBuilder unsatCoreBuilder{config, logic, proof, pmanager, termNames};
```
Comment (blishko) [inline src/api/MainSolver.cc:235]: No need for the temporary variable anymore, no?

File: src/unsatcores/MinUnsatCoreBuilder.cc (hunk 4 above)
Comment (Tomaqa) [reply src/unsatcores/MinUnsatCoreBuilder.cc]: -> #770

File: src/unsatcores/UnsatCoreBuilder.cc [hunk 6]
```diff
@@ -53,11 +101,94 @@ void UnsatCoreBuilder::computeTerms() {
...
+
+void UnsatCoreBuilder::minimizeInit() {
+    assert(terms.size() >= namedTerms.size());
+    assert(size_t(namedTerms.size()) == namedTermsIdxs.size());
+
+    setSmtSolverConfig();
+    setSmtSolver();
+}
+
+void UnsatCoreBuilder::minimizeAlgNaive() {
```
Comment (blishko) [inline src/unsatcores/UnsatCoreBuilder.cc:141]: To me it seems like it would be better if the new config and new solver were local to this method. Then, there would be no need for the `set` methods.

**Review by blishko — APPROVED** [review]
LGTM!
I would make the solver and config local to the method where they are needed. Feels like the class would simpler that way.
But I am also fine with how it is now.

File: src/unsatcores/UnsatCoreBuilder.cc (hunk 6 above)
Comment (Tomaqa) [reply src/unsatcores/UnsatCoreBuilder.cc:141]: But then it would not be usable in functions other than `minimizeAlgNaive`, and I plan to add more functions and select one of them based on the configuration.

File: src/api/MainSolver.cc (hunk 5 above)
Comment (Tomaqa) [reply src/api/MainSolver.cc:235]: Not necessarily but I still find it more readable.

File: src/unsatcores/UnsatCoreBuilder.cc (hunk 6 above)
Comment (blishko) [reply src/unsatcores/UnsatCoreBuilder.cc:141]: Then each method can have its own local solver.
The solver is not supposed to be shared between different methods, no?

Comment (blishko) [reply src/unsatcores/UnsatCoreBuilder.cc:141]: My concerns are readability and simplicity.

Comment (Tomaqa) [reply src/unsatcores/UnsatCoreBuilder.cc:141]: Another thought: minimization stuff can be enclosed in a nested class (e.g. `Minimize`) where common functionalities among various strategies can be shared using inheritance, and member variables within the class can be "local" in a sense - within the scope of minimization. Also `minimizeInit` may be replaced by constructor, `minimizeFinish` by destructor, etc. It would still not use the variables as local variables within a function, but the scope of the variables would still be limited.

Comment (blishko) [reply src/unsatcores/UnsatCoreBuilder.cc:141]: How about we merge this now and you can rethink the design when you want to add another minimization strategy?

Comment (Tomaqa) [reply src/unsatcores/UnsatCoreBuilder.cc:141]: I got a similar idea. I will make it local for now and we will see what to do next.

**Review by blishko — APPROVED** [review]
LGTM!
